### PR TITLE
feature(new integration): Pinterest Conversion API adding event multiplexing

### PR DESCRIPTION
**Pull request by shrouti1507**
## Description of the change

Adding event multiplexing for supporting duplicate mapping

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/pinterest_tag/transform.js [hunk 1]
```diff
@@ -159,9 +116,65 @@ const processEcomFields = (message, mandatoryPayload) => {
...
+  let processedUserPayload;
+  const { appId, advertiserId, deduplicationKey, sendingUnHashedData } = Config;
+  // ref: https://s.pinimg.com/ct/docs/conversions_api/dist/v3.html
+  const processedCommonPayload = processCommonPayload(message);
+  /*
+    message deduplication facility is provided *only* for the users who are using the *new configuration*.
+    if  "enableDeduplication" is set to *true* and "deduplicationKey" is set via webapp, that key value will be
+    sent as "event_id". On it's absence it will fallback to "messageId".
+    And if "enableDeduplication" is set to false, it will fallback to "messageId"
+  */
```
Comment (utsabc) [inline v0/destinations/pinterest_tag/transform.js:134]: These can be moved as part of this function's JS doc

File: v0/destinations/pinterest_tag/transform.js [hunk 2]
```diff
@@ -159,9 +116,65 @@ const processEcomFields = (message, mandatoryPayload) => {
...
+  let response = {
+    ...processedCommonPayload,
+    event_name: eventName,
+    app_id: appId,
+    advertiser_id: advertiserId,
+    user_data: processedUserPayload
+  };
+
+  if (messageType === EventType.TRACK) {
+    response = processEcomFields(message, response);
```
Comment (utsabc) [inline v0/destinations/pinterest_tag/transform.js:166]:
```suggestion
    response = postProcessEcomFields(message, response);
```

**Review by utsabc — CHANGES_REQUESTED** [review]

File: v0/destinations/pinterest_tag/utils.js [hunk 3]
```diff
@@ -159,11 +159,16 @@ const deduceTrackScreenEventName = (message, Config) => {
...
+      Config.eventsMapping,
+      "from",
+      "to",
+      false
+    );
     eventName = keyMap[trackEventOrScreenName];
   }
   if (isDefined(eventName)) {
-    return eventName;
+    return [...eventName];
```
Comment (utsabc) [inline v0/destinations/pinterest_tag/utils.js:171]: We need to treat `eventName` as array which we are not here

**Review by utsabc — DISMISSED** [review]

**Review by ItsSudip — DISMISSED** [review]

File: v0/destinations/pinterest_tag/transform.js [hunk 4]
```diff
@@ -271,12 +290,15 @@ const processRouterDest = inputs => {
...
-          metadata: event.metadata,
-          destination
-        };
-        successRespList.push(transformedPayload);
+        const transformedMessageArray = process(event);
+        transformedMessageArray.forEach(singleResponse => {
+          const transformedPayload = {
+            message: singleResponse,
+            metadata: event.metadata,
+            destination
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply v0/destinations/pinterest_tag/transform.js:298]: yes, I see 2 events with same job id.

File: v0/destinations/pinterest_tag/utils.js (hunk 3 above)
Comment (shrouti1507) [reply v0/destinations/pinterest_tag/utils.js:171]: you mean while declaration we need to do `let eventName = [];`?
`getHashFromArrayWithDuplicate` function returns a set, so in this line I am converting it to array

File: v0/destinations/pinterest_tag/transform.js (hunk 1 above)
Comment (shrouti1507) [reply v0/destinations/pinterest_tag/transform.js]: moved

File: v0/destinations/pinterest_tag/transform.js [hunk 5]
```diff
@@ -159,9 +116,65 @@ const processEcomFields = (message, mandatoryPayload) => {
...
+  }
+  // return responseBuilderSimple(response);
+  return response;
+};
+
 const process = event => {
-  let response = {};
-  let mandatoryPayload = {};
+  // let response = {};
+  // let mandatoryPayload = {};
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply v0/destinations/pinterest_tag/transform.js]: removed

File: v0/destinations/pinterest_tag/transform.js (hunk 2 above)
Comment (shrouti1507) [reply v0/destinations/pinterest_tag/transform.js]: addressed

File: v0/destinations/pinterest_tag/transform.js [hunk 6]
```diff
@@ -180,7 +118,18 @@ const process = event => {
     case EventType.PAGE:
     case EventType.SCREEN:
     case EventType.TRACK:
-      mandatoryPayload = commonFieldResponseBuilder(message, destination);
+      deducedEventNameArray = deduceEventName(message, destination.Config);
```
Comment (ItsSudip) [inline v0/destinations/pinterest_tag/transform.js:121]:
```suggestion
      deducedEventNameArray.push(
        ...deduceEventName(message, destination.Config)
      );
```

File: v0/destinations/pinterest_tag/transform.js [hunk 7]
```diff
@@ -73,95 +80,26 @@ const commonFieldResponseBuilder = (message, { Config }) => {
...
+  // return responseBuilderSimple(response);
+  return response;
 };
 
 const process = event => {
-  let response = {};
-  let mandatoryPayload = {};
+  const toSendEvents = [];
+  const respList = [];
+  let deducedEventNameArray = [];
```
Comment (ItsSudip) [inline v0/destinations/pinterest_tag/transform.js:102]:
```suggestion
  const deducedEventNameArray = [];
```

**Review by ItsSudip — CHANGES_REQUESTED** [review]

**Review by ItsSudip — APPROVED** [review]

File: v0/destinations/pinterest_tag/transform.js (hunk 6 above)
Comment (krishna2020) [reply v0/destinations/pinterest_tag/transform.js]: @ItsSudip what's the need for exploding array & again pushing it to new array ?

File: v0/destinations/pinterest_tag/transform.js [hunk 8]
```diff
@@ -73,95 +80,26 @@ const commonFieldResponseBuilder = (message, { Config }) => {
...
-  return {
+  let response = {
     ...processedCommonPayload,
-    event_name: deducedEventName,
+    event_name: eventName,
     app_id: appId,
     advertiser_id: advertiserId,
     user_data: processedUserPayload
   };
-};
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply v0/destinations/pinterest_tag/transform.js:91]: No it was just moved to utils.js file for this destination only. In order to keep the transform.js file clean

File: v0/destinations/pinterest_tag/transform.js (hunk 6 above)
Comment (ItsSudip) [reply v0/destinations/pinterest_tag/transform.js]: I just wanted to make the array constant. That is why apart from creating a new array I asked to push in the existing array.

**Review by krishna2020 — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**